contracts/RelayerHub.sol

```diff
@@ -8,88 +8,189 @@ import "./System.sol";
-    emit relayerUnRegister(msg.sender);
-  }
-
-  /*********************** Param update ********************************/
-  function updateParam(string calldata key, bytes calldata value) external override onlyInit onlyGov{
-    if (Memory.compareStrings(key,"requiredDeposit")) {
-      require(value.length == 32, "length of requiredDeposit mismatch");
-      uint256 newRequiredDeposit = BytesToTypes.bytesToUint256(32, value);
-      require(newRequiredDeposit > 1 && newRequiredDeposit <= 1e21 && newRequiredDeposit > dues, "the requiredDeposit out of range");
-      requiredDeposit = newRequiredDeposit;
-    } else if (Memory.compareStrings(key,"dues")) {
-      require(value.length == 32, "length of dues mismatch");
-      uint256 newDues = BytesToTypes.bytesToUint256(32, value);
-      require(newDues > 0 && newDues < requiredDeposit, "the dues out of range");
-      dues = newDues;
-    } else {
-      require(false, "unknown param");
-    }
-    emit paramChange(key, value);
-  }
-
-  function isRelayer(address sender) external override view returns (bool) {
-    return relayersExistMap[sender];
-  }
+contract RelayerHub is IRelayerHub, System, IParamSubscriber {
+    using SafeMath for uint256;
+
+    uint256 public constant INIT_REQUIRED_DEPOSIT = 1e20;
+    uint256 public constant INIT_DUES = 1e17;
+    address public constant WHITELIST_1 = 0xb005741528b86F5952469d80A8614591E3c5B632;
+    address public constant WHITELIST_2 = 0x446AA6E0DC65690403dF3F127750da1322941F3e;
+
+    uint256 public requiredDeposit;
+    uint256 public dues;
+
+    mapping(address => manager) managers;
+    mapping(address => bool) managersRegistered;
+    mapping(address => bool) relayManagersExistMap;
+    mapping(address => address) managersAndRelayers;
+    mapping(address => bool) relayerExistsMap;
+
+    struct manager {
+        uint256 deposit;
+        uint256 dues;
+    }
+
+    modifier onlyNonRegisteredManager() {
+        require(relayManagersExistMap[msg.sender], "manager does not exist");
+        _;
+    }
+
+    modifier onlyRegisteredManager() {
+        require(relayManagersExistMap[msg.sender], "manager does not exist");
+        require(managersRegistered[msg.sender], "manager not registered");
+        _;
+    }
+
+    modifier noProxy() {
+        require(msg.sender == tx.origin, "no proxy is allowed");
+        _;
+    }
+
+    event relayerRegister(address _relayer);
+    event relayerUnRegister(address _relayer);
+    event paramChange(string key, bytes value);
+
+    event removeManagerByGovEvent(address _removedManager);
+    event addManagerByGovEvent(address _addedManager);
+    event registerManagerEvent(address _registeredManager);
+    event addRelayerEvent(address _relayerToBeAdded);
+    event removeRelayerEvent(address _removedRelayer);
+
+
+    function init() external onlyNotInit {
+        requiredDeposit = INIT_REQUIRED_DEPOSIT;
+        dues = INIT_DUES;
+        addInitRelayer(WHITELIST_1);
```

Comment:
The `init` will never call again, so the modification here is useless, it means `addInitRelayer`  never done.

Reply:
Could you tell the exact hardfork steps with regards to the changes in this PR? I was under the impression that when we change the BSC and deploy the new RelayerHub, the `init()` can be called again. 

Also how else can we ensure smoother transition so that the current relayers remain relayers right after the hardfork? By initialising the maps in a constructor?

Reply:
Since `init()` or `construct()` can't be called again, I have [added](https://github.com/bnb-chain/bsc-genesis-contract/pull/205/commits/6dfa1690db3275431c080108c54e7571013fb988) the `update()` function which can only be called by a pre-decided address so that the code inside update() can be executed manually right after the deployment of the smart contract to ensure smooth transition. If this is acceptable then we can decide on the address. 

I also added the `exitOldRelayers()` for old relayers to call and exit.